### PR TITLE
ECOM-5248

### PR DESCRIPTION
**Pull request by awais786**
Implement [autocomplete](http://django-autocomplete-light.readthedocs.io/en/master/index.html)
Add lookups for courses , videos and orgs.
Add test cases.

### CODE REVIEW

File: course_discovery/apps/course_metadata/admin.py [hunk 1]
```diff
@@ -19,6 +19,31 @@ class PositionInline(admin.TabularInline):
     extra = 0
 
 
+class CorporateEndorsementInline(admin.TabularInline):
+    model = Program.corporate_endorsements.through
```
Comment (clintonb) [inline course_discovery/apps/course_metadata/admin.py]: This should be `CorporateEndorsement`. The same for the others below.

**Comment by awais786** [comment]
@bderusha I have added the validation. 
screen shot 2016-08-19 at 12 01 44 am

**Comment by awais786** [comment]
@bderusha Please review this PR. I will complete the tests cases tomorrow.

**Comment by bderusha** [comment]
@awais786 any updates on how feasible it would be to only populate the `excluded_course_runs` field with course_runs with those of currently saved courses?

**Comment by awais786** [comment]
@bderusha unfortunately its not working at the moment. I will try again tomorrow and let you know.

File: course_discovery/apps/course_metadata/admin.py (hunk 1 above)
Comment (awais786) [reply course_discovery/apps/course_metadata/admin.py]: @clintonb sorry for the late reply.  For many to many relation `through` is used for accessing the table.  See [Manytomany Relation](https://docs.djangoproject.com/en/dev/ref/contrib/admin/#working-with-many-to-many-models).

**Comment by awais786** [comment]
New PR created using default admin https://github.com/edx/course-discovery/pull/265

**Comment by awais786** [comment]
@waheedahmed @bderusha Please review this PR.

**Comment by bderusha** [comment]
Can we make the video dropdown show the description rather than the url?  Or both?

**Comment by bderusha** [comment]
I'm not sure that courses should be a required field to save.  I'd imagine that it is possible we would want to create the program shell before the courses exist?  @clintonb thoughts?

**Comment by clintonb** [comment]
@bderusha courses should be required to publish. I am fine with loosening restrictions for saves.

**Comment by awais786** [comment]
@bderusha in that case same restriction like banners can be added "Active status should have some courses and course runs". And allow the empty courses at the time of program creation.

**Comment by bderusha** [comment]
@awais786 Yes, I think that makes sense

File: course_discovery/apps/course_metadata/lookups.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+from .models import Course, Organization, Video
+
+
+class CourseAutocomplete(autocomplete.Select2QuerySetView):
+    def get_queryset(self):
+        if self.request.user.is_authenticated() and self.request.user.is_staff:
+            qs = Course.objects.all()
+            if self.q:
+                qs = qs.filter(key__istartswith=self.q)
```
Comment (bderusha) [inline course_discovery/apps/course_metadata/lookups.py]: can we filter on key OR title using `icontains`

File: course_discovery/apps/course_metadata/lookups.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        return []
+
+
+class OrganizationAutocomplete(autocomplete.Select2QuerySetView):
+    def get_queryset(self):
+        if self.request.user.is_authenticated() and self.request.user.is_staff:
+            qs = Organization.objects.all()
+
+            if self.q:
+                qs = qs.filter(key__istartswith=self.q)
```
Comment (bderusha) [inline course_discovery/apps/course_metadata/lookups.py]: same here

File: course_discovery/apps/course_metadata/lookups.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+        return []
+
+
+class VideoAutocomplete(autocomplete.Select2QuerySetView):
+    def get_queryset(self):
+        if self.request.user.is_authenticated() and self.request.user.is_staff:
+            qs = Video.objects.all()
+            if self.q:
+                qs = qs.filter(description__istartswith=self.q)
```
Comment (bderusha) [inline course_discovery/apps/course_metadata/lookups.py]: let's filter this on src and description using `icontains`

File: course_discovery/apps/course_metadata/lookups.py (hunk 2 above)
Comment (bderusha) [inline course_discovery/apps/course_metadata/lookups.py]: Can we also update this to make the video results populate with both src and description

**Comment by awais786** [comment]
@bderusha In this PR I have removed constraint of courses for a program. I have created a followup story. [ECOM-5385](https://openedx.atlassian.net/browse/ECOM-5385)

**Comment by bderusha** [comment]
Can we make the course_runs render the same as excluded_course_runs

screen shot 2016-08-26 at 11 29 41 am

File: course_discovery/apps/course_metadata/tests/test_admin.py [hunk 5]
```diff
@@ -153,3 +152,14 @@ def valid_post_form(self, data, file_data):
...
         self.assertEqual(response.status_code, 200)
+
+    def test_new_program_without_courses(self):
+        """ Verify that new program can be added without `courses`."""
+        data = self._post_data(Program.ProgramStatus.Unpublished)
+        data['courses'] = []
+        form = ProgramAdminForm(data)
+        self.assertTrue(form.is_valid())
+        program = form.save()
+        self.assertEqual(program.courses.all(), [])
```
Comment (bderusha) [inline course_discovery/apps/course_metadata/tests/test_admin.py]: `self.assertEqual(len(program.courses.all()), 0)`

**Comment by awais786** [comment]
@bderusha I tried that earlier. But it need some more time. I will try to find a solution again.

**Comment by awais786** [comment]
PR is ready for final review.

**Comment by bderusha** [comment]
👍

**Pull request merged**